feat(sequencer): add ttl and invalid cache to app mempool

Lilyjjo opened the pull request:
## Summary
This PR adds to the App's mempool a way to signal to CometBFT when transactions should be removed from the CometBFT mempool. 

## Background
The CometBFT and App mempools currently get transactions that fail to execute stuck in them. This is because when a transaction fails to execute in `prepare_proposal()`, it doesn't get removed from the CometBFT mempool. CometBFT only clears out transactions if they either fail `handleCheckTx()` or are included in a block. Because of this, CometBFT will re-add the failed transaction to the App mempool during its `handleCheckTx()` maintenance, which will cause it to be fed to `prepare_proposal()` again.

We also need a way for full nodes to clear out these invalid transactions. Since these nodes don't run `prepare_proposal()`, we need an additional way, like a tx TTL, to signal when transactions can be dropped.

## Changes
- Added a transaction cache to the App's mempool which signals to CometBFT when a transaction should be removed.
- Tracks in the App's mempool when a transaction is first seen.
- Transactions that fail in `prepare_proposal()` get added to the App's removal cache.
- Transactions that are older than 10 minutes are added to the App's removal cache.

## Testing
Unit tests and local testing of invalid transactions. 

## Metrics
Added `CHECK_TX_REMOVED_FAILED_EXECUTION` counter to the sequencer's metrics. 
Added `CHECK_TX_REMOVED_EXPIRED` counter to the sequencer's metrics.

## Related
Closes #979

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -62,13 +67,18 @@ impl PartialOrd for TransactionPriority {
         Self {
             tx_hash: signed_tx.sha256_of_proto_encoding(),
             signed_tx: Arc::new(signed_tx),
+            ttl: Time::now()
+                .unix_timestamp()
+                .checked_add(ttl)
+                .expect("overflow in enqueued transaction ttl"),
```
i'd prefer having a `timeout_time` passed in, which is a unix timestamp of when the tx expires, over a ttl. it'll be easier to test and compute client-side when the exact timeout is going to be before submitting

Lilyjjo replied:
would the time be decided by the transaction signer? I'm a little hesitant to let users add an expiration if they aren't signing over the expiration. future block builders wouldn't have to respect non-signed data if they didn't want to

noot replied:
okay right, since this doesn't actually add it to the tx type, this is fine

Fraser999 commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
+            let removed_tx = self
+                .remove_queue
+                .pop_front()
+                .expect("cache should contain elements");
```
I don't think we need to panic here - maybe an error log if we get `None`?

Alternatively, if we changed `max_size` to be a `std::num::NonZeroUsize` and changed `if self.size == self.max_size` to `if self.remove_queue.len() == usize::from(self.max_size)` it would be reasonable to leave the `expect` since it could never be hit.

Fraser999 commented on crates/astria-sequencer/src/service/mempool.rs:
```diff
@@ -187,6 +187,17 @@ async fn handle_check_tx<S: StateReadExt + 'static>(
+        mempool.remove(tx_hash).await;
+        metrics::counter!(metrics_init::CHECK_TX_REMOVED_INVALID).increment(1);
+        return response::CheckTx {
+            code: AbciErrorCode::INVALID_PARAMETER.into(),
```
Not sure if this is a good code - maybe we need a new one for this case?

noot replied:
yeah, a code specifically for this would be good

Lilyjjo commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -126,68 +208,107 @@ impl std::hash::Hash for EnqueuedTransaction {
     /// inserts a transaction into the mempool
     ///
-    /// note: if the tx already exists in the mempool, it's overwritten with the new priority.
+    /// note: if the tx already exists in the mempool it's a noop
```
I didn't want the original TTL to be overwritten, the `insert_all()` takes Enqueued transactions that already have a TTL set. I need to better document / assert this!

noot replied:
you could put the original ttl into the current tx and overwrite the queued value with the new priority

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
+    remove_queue: VecDeque<[u8; 32]>,
+    max_size: usize,
+    time_to_live: i64,
+    size: usize,
```
is this the current size of the cache? could you add a comment for this?

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
+        // the tx is known and entry hasn't expired
+        self.cache.contains(&tx_hash)
+            && (Time::now().unix_timestamp()
+                <= self.time_added[&tx_hash]
+                    .checked_add(self.time_to_live)
+                    .expect("overflowed ttl add"))
+    }
```
it might be nicer to store `expiry_time` instead of `time_added` when a tx is inserted, that way you don't need to do the add + expect here each time

noot replied:
or maybe have separate functions for `is_cached` and `is_expired`?

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
+            self.cache.remove(&removed_tx);
+            self.time_added.remove(&removed_tx);
+        } else {
+            self.size = self.size.checked_add(1).expect("cache size overflowed");
```
do we need a `self.size` variable over just checking `self.remove_queue.len()` or len of the hashmap?

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
+        }
+    }
+
+    fn cached(&self, tx_hash: [u8; 32]) -> bool {
```
this funtion returns `true` if the tx is invalidated but not expired right? could you add a comment for that or maybe change the function name to `cached_and_not_expired` for clarity?

noot replied:
is this the right behaviour? based on the logic for adding txs to the invalid cache and when to remove them from cometbft, should this return true if the tx is expired?

Lilyjjo replied:
the comment was ambiguous, it was returning true if it was expired in the app's mempool but not expired in the cache. refactored to make it more clear

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -126,68 +208,107 @@ impl std::hash::Hash for EnqueuedTransaction {
+    queue: Arc<RwLock<MempoolQueue>>,
+    set: Arc<RwLock<HashSet<[u8; 32]>>>,
+    invalid_cache: Arc<RwLock<InvalidCache>>,
+    tx_ttl: i64,
```
do we need to store this if it's already a constant? it appears to always be set to the same value

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -126,68 +208,107 @@ impl std::hash::Hash for EnqueuedTransaction {
+            self.set.write().await.remove(&tx.tx_hash);
+            Some((tx, priority))
+        } else {
+            tx
```
```suggestion
            None
```

a bit more explicit

Lilyjjo commented:
Changes made since last reviews:
- switched time accounting to `tokio::time::{Instant, Duration}`
- renamed `InvalidCache` to `RemovalCache` and renamed Mempool’s use of it to `comet_bft_removal_cache` to try to better document what the cache is and how it is being used 
- simplified the `RemovalCache` by removing the redundant `HashSet` and `size` struct members
- removed the `HashSet` added to the mempool in the last write
- added tracking to `RemovalCache` as to why the transaction was added and surface error in CheckTx to user
- added `AbciErrorCode`s and metrics for expired transactions and failing to execute transactions

noot commented on crates/astria-sequencer/Cargo.toml:
```diff
@@ -50,7 +50,7 @@ serde_json = { workspace = true }
 tendermint-proto = { workspace = true }
 tendermint = { workspace = true }
-tokio = { workspace = true, features = ["rt", "tracing"] }
+tokio = { workspace = true, features = ["rt", "tracing", "test-util"] }
```
if the `test-util` feature is just for tests, move it under `dev-dependencies`

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +133,81 @@ impl std::hash::Hash for EnqueuedTransaction {
+}
+
+const TX_TTL: Duration = Duration::from_secs(600); // 10 minutes 
+const REMOVAL_CACHE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(4096) }; // TODO make configuration variable 
```
why not just use `usize`?

Lilyjjo replied:
Fraser recommended this instead I'm guessing for configuration safety purposes, but I can change back to just a `usize`

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +133,81 @@ impl std::hash::Hash for EnqueuedTransaction {
+const REMOVAL_CACHE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(4096) }; // TODO make configuration variable 
+const TX_REMOVAL_TTL: Duration = Duration::from_secs(30); // 30 seconds
+
+/// `RemovalCache` is a cache used for signaling to `CometBFT` when a
+/// transaction can be removed from the `CometBFT` mempool outside of the
+/// `CheckTx` checks.
```
this is checked in `check_tx` right? there isn't a way to tell cometbft to remove something from the mempool without cometbft calling `check_tx` on it

Lilyjjo replied:
My b, yeah the comment should read 'other CheckTX checks'

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +133,81 @@ impl std::hash::Hash for EnqueuedTransaction {
+/// that failed to execute or to re-add a transaction that has expired.
+#[derive(Clone)]
+pub(crate) struct RemovalCache {
+    cache: HashMap<[u8; 32], (Instant, Arc<anyhow::Error>)>,
```
for the error, the only two options right now are expired and failed execution right? i'd prefer to change the `Arc<anyhow::Error>` to an enum, that would be clearer as to which removal reasons exist

Lilyjjo replied:
I was trying to bubble up the reason for failure to the user, but it gets stuck at 'failed execution' or 'failed stateful checks'. I can make it an enum if you don't think that the `CheckTx` failure case should be more specific than 'failed execution'

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +133,81 @@ impl std::hash::Hash for EnqueuedTransaction {
+    }
+
+    /// returns Some(Error) if transaction is cached, where the Error is
+    /// the reason why it was cached. None is returned if the transaction
+    /// is not cached
```
```suggestion
    /// the reason why it was cached. None is returned if the transaction
    /// is not cached or is not expired
```

Lilyjjo replied:
I was considering a tx being 'cached' as being both not expired and in the cache, but I can change

noot replied:
i see, yeah i got confused from the variable naming, a comment on that would help!

noot replied:
sorry i got confused about the tx removal ttl vs the tx in mempool ttl. this makes sense!

noot commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -157,23 +255,33 @@ impl Mempool {
     /// inserts all the given transactions into the mempool
     pub(crate) async fn insert_all(&self, txs: Vec<(EnqueuedTransaction, TransactionPriority)>) {
-        self.inner.write().await.extend(txs);
+        self.queue.write().await.extend(txs);
```
this isn't the same behaviour as `insert` as it doesn't preserve the timestamp if it already exists

Lilyjjo replied:
You're right. I was just relying on the behavior that we're currently only using this function to re-insert transactions that we just popped. Do you have a preference of me adding a comment that explains that or to do a similar process as in `insert()`?

noot replied:
i think doing a similar process as to `insert()` makes the most sense!

Lilyjjo commented:
Logic changes since last review:
- made `insert_all()` preserve timestamp like `insert()`
- made RemovalCache errors into enum

Lilyjjo commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +130,98 @@ impl std::hash::Hash for EnqueuedTransaction {
 }
 
+/// inserts or updates the transaction in a timestamp preserving manner
+fn update_or_insert(
```
@Fraser999 is this an okay place to put a function like this? The rust CLI suggested I move it out of the Mempool struct's implementation because it wasn't using the `self` variable

Fraser999 replied:
Absolutely.  Another choice would be to leave it inside the `impl Mempool` but just don't have `self` as a parameter.  You have to call it like `Self::update_or_insert` in that case.  It can make sense to do that in cases like this where you have more than struct in scope which _could_ use this function, but it's really only intended to be used by the `Mempool` struct.  But for a private function like this, it's pretty much down to personal preference :)

You _could_ simplify the signature though by just taking `queue: &mut PriorityQueue` and at callsites passing `&mut *self.queue.write().await`.  Definitely not a big deal though :)

Fraser999 commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -72,6 +84,7 @@ impl EnqueuedTransaction {
             tx_hash: signed_tx.sha256_of_proto_encoding(),
             signed_tx: Arc::new(signed_tx),
             address,
+            time_first_seen: Instant::now(),
```
This is probably better suited to live in the `TransactionPriority` struct since it might be considered in the future as part of the prioritisation algo, and also that would keep `EnqueuedTransaction`'s hash a genuine way to determine equality/ordering.

Lilyjjo commented on crates/astria-sequencer/src/mempool.rs:
```diff
@@ -117,6 +130,98 @@ impl std::hash::Hash for EnqueuedTransaction {
+            tx_hash = %telemetry::display::hex(&tx_hash),
+            "inserted transaction into mempool"
+        );
+    }
```
rewrote in moving the `time_first_seen` into the `TransactionPriority` struct

Lilyjjo commented:
Changes since last reviews:
- Removed TTL from `RemovalCache`.
- Moved `time_first_seen` out of the `EnqueuedTransaction` struct into the `TransactionPriority` struct.
